src/lib/core/common-behaviors/id.ts

```diff
@@ -0,0 +1,35 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Constructor} from './constructor';
+
+/** Monotonically increasing integer used to auto-generate unique ids. */
+let nextUniqueId = 0;
+
+/** @docs-private */
+export interface CanId {
+  id: string;
+}
+
+/**
+ * Mixin to augment a directive with an `id` property.
+ * The id property will fall back to auto-generated unique ids if developers don't specify one.
+ */
+export function mixinId<T extends Constructor<{}>>(base: T, componentName: string)
```

Comment:
Yeah. Not sure if it would make sense for them to even fall back to id's that include the `md-checkbox` prefix for example.

I think it makes sense to throw an error if the id is set `null`. Anything else before I make the changes?